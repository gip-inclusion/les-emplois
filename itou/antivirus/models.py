from django.db import models
from django.utils.text import capfirst

from itou.files.models import File


class Scan(models.Model):
    file = models.OneToOneField(File, on_delete=models.CASCADE)
    clamav_signature = models.TextField()
    clamav_completed_at = models.DateTimeField(null=True, verbose_name="analyse ClamAV le")
    clamav_infected = models.BooleanField(null=True, verbose_name="fichier infect√© selon ClamAV")
    comment = models.TextField(blank=True, verbose_name="commentaire")

    class Meta:
        verbose_name = "analyse antivirus"
        verbose_name_plural = "analyses antivirus"

    def __str__(self):
        text = f"{capfirst(self._meta.verbose_name)} {self.file_id}"
        if self.clamav_infected:
            text = f"[VIRUS] {text}"
        return text
