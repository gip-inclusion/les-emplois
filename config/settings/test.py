import os


ITOU_ENVIRONMENT = "DEV"
os.environ["ITOU_ENVIRONMENT"] = ITOU_ENVIRONMENT

from .base import *  # noqa: E402,F403


SECRET_KEY = "foobar"
ALLOWED_HOSTS = []

# We *want* to do the same `collectstatic` on the CI than on PROD to catch errors early,
# but we don't want to do it when running the test suite locally for performance reasons.
if not os.getenv("CI", False):
    # `ManifestStaticFilesStorage` (used in base settings) requires `collectstatic` to be run.
    STORAGES["staticfiles"]["BACKEND"] = "django.contrib.staticfiles.storage.StaticFilesStorage"  # noqa: F405

ASP_ITOU_PREFIX = "XXXXX"  # same as in our fixtures
ITOU_PROTOCOL = "http"
ITOU_FQDN = "localhost:8000"

DATABASES["default"]["HOST"] = os.getenv("PGHOST", "127.0.0.1")  # noqa: F405
DATABASES["default"]["PORT"] = os.getenv("PGPORT", "5432")  # noqa: F405
DATABASES["default"]["NAME"] = os.getenv("PGDATABASE", "itou")  # noqa: F405
DATABASES["default"]["USER"] = os.getenv("PGUSER", "postgres")  # noqa: F405
DATABASES["default"]["PASSWORD"] = os.getenv("PGPASSWORD", "password")  # noqa: F405

MAILJET_API_KEY_PRINCIPAL = "API_MAILJET_KEY_PRINCIPAL"
MAILJET_SECRET_KEY_PRINCIPAL = "API_MAILJET_SECRET_PRINCIPAL"

AWS_S3_ENDPOINT_URL = f"http://{os.getenv('CELLAR_ADDON_HOST', 'localhost:9000')}/"
AWS_S3_ACCESS_KEY_ID = "minioadmin"
AWS_S3_SECRET_ACCESS_KEY = "minioadmin"
AWS_STORAGE_BUCKET_NAME = "tests"
