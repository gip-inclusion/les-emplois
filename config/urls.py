from django.conf import settings
from django.contrib import admin
from django.urls import include, path, re_path, register_converter
from django.views.generic import TemplateView

from itou.utils.urls import SiretConverter
from itou.www.dashboard import views as dashboard_views
from itou.www.login import views as login_views
from itou.www.signup import views as signup_views


register_converter(SiretConverter, "siret")

urlpatterns = [
    path("admin/", admin.site.urls),
    # --------------------------------------------------------------------------------------
    # allauth URLs. Order is important because some URLs are overriden.
    # --------------------------------------------------------------------------------------
    # Override allauth `account_signup` URL.
    # /accounts/signup/ <=> account_signup
    # We don't want any user to be able to signup using the default allauth `signup` url
    # because we have multiple specific signup processes for different kind of users.
    re_path(r"^accounts/signup/$", signup_views.signup),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_login` URL.
    # /accounts/login/ <=> account_login
    # We override this view because the login page should look slightly differently
    # for job seekers, prescribers and employers.
    # Also, PEAMU is only available for job seekers.
    re_path(r"^accounts/login/$", login_views.login),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_change_password` URL.
    # /accounts/password/change/ <=> account_change_password
    # https://github.com/pennersr/django-allauth/issues/468
    re_path(r"^accounts/password/change/$", dashboard_views.password_change),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_reset_password` URL.
    # Avoid user enumeration via password reset page.
    re_path(r"^accounts/password/reset/$", signup_views.ItouPasswordResetView.as_view()),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_logout` URL.
    # /accounts/logout/ <=> account_logout
    # We need custom code to process PEAMU logout.
    re_path(r"^accounts/logout/$", dashboard_views.logout),
    # --------------------------------------------------------------------------------------
    # Other allauth URLs.
    path("accounts/", include("allauth.urls")),
    # --------------------------------------------------------------------------------------
    # PEAMU URLs.
    path("accounts/", include("itou.allauth_adapters.peamu.urls")),
    # FranceConnect URLs.
    path("franceconnect/", include("itou.france_connect.urls")),
    # --------------------------------------------------------------------------------------
    # API.
    path("api/v1/", include("itou.api.urls", namespace="v1")),
    path("api/", include("itou.www.api.urls")),
    # www.
    path("", include("itou.www.home.urls")),
    path("apply/", include("itou.www.apply.urls")),
    path("approvals/", include("itou.www.approvals_views.urls")),
    path("autocomplete/", include("itou.www.autocomplete.urls")),
    path("dashboard/", include("itou.www.dashboard.urls")),
    path("employee_record/", include("itou.www.employee_record_views.urls")),
    path("institutions/", include("itou.www.institutions_views.urls")),
    path("invitations/", include("itou.www.invitations_views.urls")),
    path("prescribers/", include("itou.www.prescribers_views.urls")),
    path("search/", include("itou.www.search.urls")),
    path("siae/", include("itou.www.siaes_views.urls")),
    path("signup/", include("itou.www.signup.urls")),
    path("stats/", include("itou.www.stats.urls")),
    path("versions/", include("itou.www.releases.urls")),
    # Enable Anymailâ€™s status tracking
    # https://anymail.readthedocs.io/en/stable/esps/mailjet/#status-tracking-webhooks
    re_path(r"^webhooks/anymail/", include("anymail.urls")),
    path("welcoming_tour/", include("itou.www.welcoming_tour.urls")),
    # Static pages.
    path("accessibility/", TemplateView.as_view(template_name="static/accessibility.html"), name="accessibility"),
    # Errors pages.
    path("403/", TemplateView.as_view(template_name="403.html"), name="403"),
    path("404/", TemplateView.as_view(template_name="404.html"), name="404"),
    path("500/", TemplateView.as_view(template_name="500.html"), name="500"),
]

if settings.DEBUG and "debug_toolbar" in settings.INSTALLED_APPS:
    import debug_toolbar

    urlpatterns = [path("__debug__/", include(debug_toolbar.urls))] + urlpatterns
    # In the local developpemnt environment, we enable the default callback URLs
    urlpatterns = [path("", include("itou.france_connect.urls"))] + urlpatterns
