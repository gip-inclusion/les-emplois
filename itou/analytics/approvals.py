from django.db.models import Count, Q

from itou.approvals import models as approvals_models
from itou.utils.apis import enums as api_enums

from . import models


def collect_approvals_counts(before):
    counts = approvals_models.Approval.objects.filter(created_at__lt=before).aggregate(
        total=Count("pk"),
        pe_notify_success=Count("pk", filter=Q(pe_notification_status=api_enums.PEApiNotificationStatus.SUCCESS)),
        pe_notify_pending=Count(
            "pk",
            filter=Q(
                pe_notification_status__in=(
                    api_enums.PEApiNotificationStatus.PENDING,
                    api_enums.PEApiNotificationStatus.SHOULD_RETRY,
                )
            ),
        ),
        pe_notify_error=Count("pk", filter=Q(pe_notification_status=api_enums.PEApiNotificationStatus.ERROR)),
    )
    return {
        models.DatumCode.APPROVAL_COUNT: counts["total"],
        models.DatumCode.APPROVAL_PE_NOTIFY_SUCCESS: counts["pe_notify_success"],
        models.DatumCode.APPROVAL_PE_NOTIFY_PENDING: counts["pe_notify_pending"],
        models.DatumCode.APPROVAL_PE_NOTIFY_ERROR: counts["pe_notify_error"],
    }


def collect_cancelled_approvals(before):
    cancelled_count = approvals_models.CancelledApproval.objects.filter(created_at__lt=before).count()
    return {models.DatumCode.APPROVAL_CANCELLED: cancelled_count}


def collect_analytics_data(before):
    return {
        **collect_approvals_counts(before),
        **collect_cancelled_approvals(before),
    }
