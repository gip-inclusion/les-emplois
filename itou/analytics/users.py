from django.db.models import Count

from itou.users import enums as users_enums, models as users_models

from . import models


def collect_analytics_data(before):
    counts = {
        info["kind"]: info["count"]
        for info in users_models.User.objects.values("kind").annotate(count=Count("pk")).values("kind", "count")
    }
    return {
        models.DatumCode.USER_COUNT: sum(counts.values()),
        models.DatumCode.USER_JOB_SEEKER_COUNT: counts.get(users_enums.KIND_JOB_SEEKER, 0),
        models.DatumCode.USER_PRESCRIBER_COUNT: counts.get(users_enums.KIND_PRESCRIBER, 0),
        models.DatumCode.USER_EMPLOYER_COUNT: counts.get(users_enums.KIND_EMPLOYER, 0),
        models.DatumCode.USER_LABOR_INSPECTOR_COUNT: counts.get(users_enums.KIND_LABOR_INSPECTOR, 0),
        models.DatumCode.USER_ITOU_STAFF_COUNT: counts.get(users_enums.KIND_ITOU_STAFF, 0),
    }
