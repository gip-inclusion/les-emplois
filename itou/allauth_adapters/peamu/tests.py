# Heavily inspired by
# https://github.com/pennersr/django-allauth/blob/master/allauth/socialaccount/providers/google/tests.py
from unittest import mock

from allauth.account.models import EmailAddress, EmailConfirmation
from allauth.account.signals import user_signed_up
from allauth.socialaccount.models import SocialAccount
from allauth.socialaccount.tests import OAuth2TestsMixin
from allauth.tests import MockedResponse, TestCase
from django.core import mail
from django.test import override_settings

from itou.allauth_adapters.peamu.provider import PEAMUProvider
from itou.users import enums as users_enums
from itou.users.models import User


@override_settings(
    API_ESD={"BASE_URL": "https://some.auth.domain", "AUTH_BASE_URL": "https://some-authentication-domain.fr"}
)  # noqa
class PEAMUTests(OAuth2TestsMixin, TestCase):
    provider_id = PEAMUProvider.id

    def get_mocked_response(self, given_name="John", family_name="Doe", email="john.doe@example.com"):
        return MockedResponse(
            200,
            f"""
            {{
                "email": "{email}",
                "gender": "male",
                "given_name": "{given_name}",
                "family_name": "{family_name}",
                "sub": "108204268033311374519"
            }}
            """,
        )

    @mock.patch("itou.external_data.signals.import_user_pe_data_on_peamu_login")
    def test_data_created_by_peamu_login(self, mock_login_signal):
        test_email = "john.doe@example.com"
        self.login(self.get_mocked_response())
        self.assertEqual(mock_login_signal.call_count, 1)
        email_address = EmailAddress.objects.get(email=test_email)
        account = email_address.user.socialaccount_set.get()
        self.assertEqual(account.extra_data["email"], "john.doe@example.com")
        self.assertEqual(account.extra_data["gender"], "male")
        self.assertEqual(account.extra_data["given_name"], "John")
        self.assertEqual(account.extra_data["family_name"], "Doe")
        # This mysterious 'testac' value actually comes from
        # https://github.com/pennersr/django-allauth/blob/master/allauth/socialaccount/tests.py#L150
        self.assertEqual(account.extra_data["id_token"], "testac")
        self.assertEqual(account.extra_data["sub"], "108204268033311374519")
        self.assertEqual(email_address.user.email, "john.doe@example.com")
        self.assertEqual(email_address.user.first_name, "John")
        self.assertEqual(email_address.user.last_name, "Doe")
        self.assertEqual(email_address.user.username, "john")
        self.assertEqual(email_address.user.is_active, True)
        self.assertEqual(email_address.user.identity_provider, users_enums.IdentityProvider.PE_CONNECT)
        # Note that a PEAMU user is automatically set as a job seeker.
        self.assertEqual(email_address.user.is_job_seeker, True)
        self.assertEqual(email_address.user.is_prescriber, False)
        self.assertEqual(email_address.user.is_siae_staff, False)
        self.assertEqual(email_address.user.is_staff, False)

    @mock.patch("itou.external_data.signals.import_user_pe_data_on_peamu_login")
    def test_peamu_connection_for_existing_non_peamu_user(self, mock_login_signal):
        email = "user@example.com"
        user = User.objects.create(username="user", is_active=True, email=email, is_job_seeker=True)
        user.set_password("test")
        user.save()
        EmailAddress.objects.create(user=user, email=email, primary=True)
        self.assertFalse(SocialAccount.objects.filter(user=user, provider=PEAMUProvider.id).exists())
        self.client.login(username=user.username, password="test")
        self.assertEqual(mock_login_signal.call_count, 0)

        self.login(self.get_mocked_response(), process="connect")
        # FIXME Seems strange to me
        self.assertEqual(mock_login_signal.call_count, 0)

        self.assertTrue(SocialAccount.objects.filter(user=user, provider=PEAMUProvider.id).exists())
        self.assertEqual(EmailAddress.objects.filter(user=user).count(), 1)
        # User email is not updated.
        self.assertEqual(EmailAddress.objects.filter(email=email).count(), 1)
        self.assertEqual(EmailAddress.objects.filter(email="john.doe@example.com").count(), 0)

    @mock.patch("itou.external_data.signals.import_user_pe_data_on_peamu_login")
    def test_email_verification_is_skipped_for_peamu_account(self, mock_login_signal):
        test_email = "john.doe@example.com"
        self.login(self.get_mocked_response())
        self.assertEqual(mock_login_signal.call_count, 1)

        email_address = EmailAddress.objects.get(email=test_email)
        self.assertFalse(email_address.verified)
        self.assertFalse(EmailConfirmation.objects.filter(email_address__email=test_email).exists())
        self.assertEqual(len(mail.outbox), 0)

    @mock.patch("itou.external_data.signals.import_user_pe_data_on_peamu_login")
    def test_username_is_based_on_first_name(self, mock_login_signal):
        first_name = "jessica"
        last_name = "parker"
        email = "john.doe@example.com"
        self.login(self.get_mocked_response(email=email, given_name=first_name, family_name=last_name))
        self.assertEqual(mock_login_signal.call_count, 1)
        user = User.objects.get(email=email)
        self.assertEqual(user.username, "jessica")

    @mock.patch("itou.external_data.signals.import_user_pe_data_on_peamu_login")
    def test_username_is_based_on_email_if_first_name_is_exotic(self, mock_login_signal):
        first_name = "明"
        last_name = "小"
        email = "john.doe@example.com"
        self.login(self.get_mocked_response(email=email, given_name=first_name, family_name=last_name))
        self.assertEqual(mock_login_signal.call_count, 1)
        user = User.objects.get(email=email)
        self.assertEqual(user.username, "john.doe")

    @mock.patch("itou.external_data.signals.import_user_pe_data_on_peamu_login")
    def test_user_signed_up_signal(self, mock_login_signal):
        sent_signals = []

        def on_signed_up(sender, request, user, **kwargs):
            sociallogin = kwargs["sociallogin"]
            self.assertEqual(sociallogin.account.provider, PEAMUProvider.id)
            self.assertEqual(sociallogin.account.user, user)
            sent_signals.append(sender)

        user_signed_up.connect(on_signed_up)
        self.login(self.get_mocked_response())
        self.assertEqual(mock_login_signal.call_count, 1)
        self.assertTrue(len(sent_signals) > 0)
