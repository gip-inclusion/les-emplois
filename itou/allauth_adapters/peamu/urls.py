from django.urls import include, re_path

from itou.allauth_adapters.peamu import views as peamu_views
from itou.allauth_adapters.peamu.provider import PEAMUProvider


def default_urlpatterns(provider):
    urlpatterns = [
        re_path(r"^login/$", peamu_views.oauth2_login, name=provider.id + "_login"),
        re_path(r"^login/callback/$", peamu_views.oauth2_callback, name=provider.id + "_callback"),
    ]

    return [
        re_path(r"^profile/$", peamu_views.redirect_to_dashboard_view),
        re_path(r"^" + provider.get_slug() + r"/", include(urlpatterns)),
    ]


urlpatterns = default_urlpatterns(PEAMUProvider)
