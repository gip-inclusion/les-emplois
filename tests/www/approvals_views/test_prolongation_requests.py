import factory
import httpx
import pytest
from django.contrib import messages
from django.core.files.storage import default_storage
from django.template import loader
from django.urls import reverse
from freezegun import freeze_time
from pytest_django.asserts import assertMessages, assertNumQueries, assertRedirects

from itou.approvals.enums import (
    ProlongationReason,
    ProlongationRequestDenyProposedAction,
    ProlongationRequestDenyReason,
    ProlongationRequestStatus,
)
from itou.files.models import File
from tests.approvals import factories as approvals_factories
from tests.prescribers import factories as prescribers_factories
from tests.users import factories as users_factories
from tests.utils.test import BASE_NUM_QUERIES, parse_response_to_soup


@pytest.mark.parametrize(
    "authorized_organization,expected",
    [
        (False, 404),
        (True, 200),
    ],
)
def test_list_view_access(client, authorized_organization, expected):
    client.force_login(users_factories.PrescriberFactory(membership__organization__authorized=authorized_organization))

    response = client.get(reverse("approvals:prolongation_requests_list"))
    assert response.status_code == expected


def test_empty_list_view(snapshot, client):
    prescriber = prescribers_factories.PrescriberOrganizationWithMembershipFactory(authorized=True).members.first()
    client.force_login(prescriber)

    num_queries = (
        BASE_NUM_QUERIES
        + 1  # fetch django session
        + 1  # fetch user
        + 1  # check user memberships
        + 1  # count prolongation requests rows (from pager)
        + 1  # `is_prescriber_with_authorized_org()` in nav
        + 3  # savepoint, update session, release savepoint
    )
    with assertNumQueries(num_queries):
        response = client.get(reverse("approvals:prolongation_requests_list"))
    assert str(parse_response_to_soup(response, ".s-section .c-box")) == snapshot


def test_list_view(snapshot, client):
    prolongation_request = approvals_factories.ProlongationRequestFactory(for_snapshot=True)
    client.force_login(prolongation_request.validated_by)

    num_queries = (
        BASE_NUM_QUERIES
        + 1  # fetch django session
        + 1  # fetch user
        + 1  # check user memberships
        + 1  # fetch prolongation requests rows
        + 1  # count prolongation requests rows (from pager)
        + 1  # `is_prescriber_with_authorized_org()` in nav
        + 3  # savepoint, update session, release savepoint
    )
    with assertNumQueries(num_queries):
        response = client.get(reverse("approvals:prolongation_requests_list"))
    assert str(parse_response_to_soup(response, ".s-section .c-box")) == snapshot


def test_list_view_only_pending_filter(client):
    pending_prolongation_request = approvals_factories.ProlongationRequestFactory(
        status=ProlongationRequestStatus.PENDING,
    )
    other_prolongation_requests = approvals_factories.ProlongationRequestFactory.create_batch(
        2,
        status=factory.Iterator([ProlongationRequestStatus.GRANTED, ProlongationRequestStatus.DENIED]),
        prescriber_organization=pending_prolongation_request.prescriber_organization,
    )
    client.force_login(pending_prolongation_request.validated_by)

    response = client.get(reverse("approvals:prolongation_requests_list"))
    assert set(response.context["pager"].object_list) == {
        pending_prolongation_request,
        *other_prolongation_requests,
    }

    response = client.get(reverse("approvals:prolongation_requests_list"), data={"only_pending": True})
    assert list(response.context["pager"].object_list) == [pending_prolongation_request]


def test_show_view_access(client):
    prolongation_request, other_prolongation_request = approvals_factories.ProlongationRequestFactory.create_batch(2)

    # When we are not yet connected
    url = reverse("approvals:prolongation_request_show", kwargs={"prolongation_request_id": prolongation_request.pk})
    assertRedirects(client.get(url), reverse("account_login") + f"?next={url}")

    client.force_login(prolongation_request.validated_by)
    # When the prolongation request is for the current prescriber organization
    response = client.get(
        reverse("approvals:prolongation_request_show", kwargs={"prolongation_request_id": prolongation_request.pk})
    )
    assert response.status_code == 200

    # When the prolongation request is for another prescriber organization
    response = client.get(
        reverse(
            "approvals:prolongation_request_show", kwargs={"prolongation_request_id": other_prolongation_request.pk}
        )
    )
    assert response.status_code == 404

    # When the prolongation request doesn't exists
    response = client.get(reverse("approvals:prolongation_request_show", kwargs={"prolongation_request_id": 0}))
    assert response.status_code == 404


def test_show_view(snapshot, client):
    prolongation_request = approvals_factories.ProlongationRequestFactory(for_snapshot=True)
    client.force_login(prolongation_request.validated_by)

    default_storage.location = "snapshot"
    response = client.get(
        reverse("approvals:prolongation_request_show", kwargs={"prolongation_request_id": prolongation_request.pk})
    )
    assert str(parse_response_to_soup(response, ".s-section .col-lg-8 .c-box:last-child")) == snapshot


def test_grant_view(client):
    prolongation_request = approvals_factories.ProlongationRequestFactory(approval__user__for_snapshot=True)
    client.force_login(prolongation_request.validated_by)

    response = client.post(
        reverse("approvals:prolongation_request_grant", kwargs={"prolongation_request_id": prolongation_request.pk}),
    )
    assertRedirects(response, reverse("approvals:prolongation_requests_list"), fetch_redirect_response=False)
    assertMessages(
        response,
        [messages.Message(messages.SUCCESS, "La prolongation de John DOE a bien été acceptée.")],
    )
    prolongation_request.refresh_from_db()
    assert prolongation_request.status == ProlongationRequestStatus.GRANTED


def test_grant_view_is_not_accessible_by_get_method(client):
    prolongation_request = approvals_factories.ProlongationRequestFactory()
    client.force_login(prolongation_request.validated_by)

    response = client.get(
        reverse("approvals:prolongation_request_grant", kwargs={"prolongation_request_id": prolongation_request.pk})
    )
    assert response.status_code == 405


@pytest.mark.parametrize("reason", ProlongationRequestDenyReason)
def test_deny_view_for_reasons(snapshot, client, reason):
    prolongation_request = approvals_factories.ProlongationRequestFactory(for_snapshot=True)
    client.force_login(prolongation_request.validated_by)

    # Reverse all needed URL
    start_url = reverse(
        "approvals:prolongation_request_deny", kwargs={"prolongation_request_id": prolongation_request.pk}
    )
    reason_url = reverse(
        "approvals:prolongation_request_deny",
        kwargs={"prolongation_request_id": prolongation_request.pk, "step": "reason"},
    )
    reason_explanation_url = reverse(
        "approvals:prolongation_request_deny",
        kwargs={"prolongation_request_id": prolongation_request.pk, "step": "reason_explanation"},
    )
    proposed_actions_url = reverse(
        "approvals:prolongation_request_deny",
        kwargs={"prolongation_request_id": prolongation_request.pk, "step": "proposed_actions"},
    )
    end_url = reverse("approvals:prolongation_requests_list")

    # Starting the tunnel should redirect to the first step
    assertRedirects(client.get(start_url), reason_url)
    # Checking the title at least once
    assert str(parse_response_to_soup(client.get(reason_url), selector="#main .s-title-01")) == snapshot(name="title")

    # Submit data for the "reason" step
    assert str(parse_response_to_soup(client.get(reason_url), selector="#main .s-section")) == snapshot(name="reason")
    response = client.post(
        reason_url,
        {"reason-reason": reason, "prolongation_request_deny_view-current_step": "reason"},
    )
    assertRedirects(response, reason_explanation_url)

    # Submit data for the "reason_explanation" step
    assert str(parse_response_to_soup(client.get(reason_explanation_url), selector="#main .s-section")) == snapshot(
        name="reason_explanation"
    )
    response = client.post(
        reason_explanation_url,
        {
            "reason_explanation-reason_explanation": "Lorem ipsum",
            "prolongation_request_deny_view-current_step": "reason_explanation",
        },
        follow=True,
    )

    if reason is ProlongationRequestDenyReason.IAE:
        assertRedirects(response, proposed_actions_url)
        assert str(parse_response_to_soup(client.get(proposed_actions_url), selector="#main .s-section")) == snapshot(
            name="proposed_actions"
        )
        # Submit data for the "proposed_actions" step
        response = client.post(
            proposed_actions_url,
            {
                "proposed_actions-proposed_actions": list(ProlongationRequestDenyProposedAction),
                "proposed_actions-proposed_actions_explanation": "Lorem ipsum",
                "prolongation_request_deny_view-current_step": "proposed_actions",
            },
            follow=True,  # formtools will redirect to "done" step to end the tunnel, then we redirect to another URL
        )

    assertRedirects(response, end_url)
    assertMessages(response, [messages.Message(messages.SUCCESS, "La prolongation de John DOE a bien été refusée.")])
    prolongation_request.refresh_from_db()
    assert prolongation_request.status == ProlongationRequestStatus.DENIED
    assert prolongation_request.deny_information.reason == reason


@pytest.mark.parametrize("status", ProlongationRequestStatus)
def test_template_status_card(snapshot, status):
    prolongation_request = approvals_factories.ProlongationRequestFactory(
        for_snapshot=True,
        processed=True,
        status=status,
    )

    assert (
        loader.render_to_string(
            "approvals/prolongation_requests/_status_box.html",
            context={
                "ProlongationRequestStatus": ProlongationRequestStatus,
                "prolongation_request": prolongation_request,
            },
        )
        == snapshot
    )


class TestProlongationReportFileView:
    def test_anonymous(self, client):
        url = reverse("approvals:prolongation_request_report_file", kwargs={"prolongation_request_id": 0})
        response = client.get(url)
        assertRedirects(response, reverse("account_login") + f"?next={url}")

    def test_nonexistent(self, client):
        prescriber = users_factories.PrescriberFactory(membership__organization__authorized=True)
        client.force_login(prescriber)
        response = client.get(
            reverse("approvals:prolongation_request_report_file", kwargs={"prolongation_request_id": 0})
        )
        assert response.status_code == 404

    def test_other_organization(self, client):
        prescriber = users_factories.PrescriberFactory(membership__organization__authorized=True)
        request = approvals_factories.ProlongationRequestFactory()
        client.force_login(prescriber)
        response = client.get(
            reverse(
                "approvals:prolongation_request_report_file",
                kwargs={"prolongation_request_id": request.pk},
            )
        )
        assert response.status_code == 404

    def test_no_report_file(self, client):
        org = prescribers_factories.PrescriberOrganizationFactory(authorized=True)
        prescriber = users_factories.PrescriberFactory(membership__organization=org)
        request = approvals_factories.ProlongationRequestFactory(prescriber_organization=org, report_file=None)
        client.force_login(prescriber)
        response = client.get(
            reverse(
                "approvals:prolongation_request_report_file",
                kwargs={"prolongation_request_id": request.pk},
            )
        )
        assert response.status_code == 404

    def test_ok(self, client, xlsx_file):
        org = prescribers_factories.PrescriberOrganizationFactory(authorized=True)
        prescriber = users_factories.PrescriberFactory(membership__organization=org)
        key = default_storage.save("prolongation_report/empty.xlsx", xlsx_file)
        file = File.objects.create(key=key)
        request = approvals_factories.ProlongationRequestFactory(
            prescriber_organization=org, reason=ProlongationReason.RQTH, report_file=file
        )
        client.force_login(prescriber)
        # Boto3 signed requests depend on the current date, with a second resolution.
        # See X-Amz-Date in
        # https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html
        with freeze_time():
            response = client.get(
                reverse(
                    "approvals:prolongation_request_report_file",
                    kwargs={"prolongation_request_id": request.pk},
                )
            )
            assertRedirects(response, default_storage.url(file.pk), fetch_redirect_response=False)
        xlsx_file.seek(0)
        assert httpx.get(response.url).content == xlsx_file.read()
