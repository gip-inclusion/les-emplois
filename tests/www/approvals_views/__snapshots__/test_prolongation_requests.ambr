# serializer version: 1
# name: test_list_view
  '''
  <div class="c-box">
                          <div class="d-flex flex-column flex-md-row align-items-md-center justify-content-md-between">
                              <h3 class="mb-0">1 résultat</h3>
                          </div>
                          <div class="d-flex flex-column flex-md-row mt-3 mt-md-4">
                              <h4 class="mr-4">Filtres :</h4>
                              <form action="/approvals/prolongation/requests" hx-boost="true" hx-trigger="change">
                                  <input id="id_only_pending" name="only_pending" type="checkbox"/>
                                  <label for="id_only_pending">Voir uniquement les demandes à traiter</label>
                              </form>
                          </div>
  
                          <div class="table-responsive-lg mt-3 mt-md-4">
                              <table class="table">
                                  <caption class="sr-only">Liste des prolongations de PASS IAE</caption>
                                  <thead>
                                      <tr>
                                          <th scope="col">#</th>
                                          <th scope="col">Salarié</th>
                                          <th scope="col">Numéro de PASS IAE</th>
                                          <th scope="col">Structure</th>
                                          <th scope="col">Demandée le</th>
                                          <th scope="col">Statut</th>
                                          <th scope="col">Action</th>
                                      </tr>
                                  </thead>
                                  <tbody>
                                      
                                          <tr>
                                              <th scope="row">1</th>
                                              <td class="font-weight-bold">John Doe</td>
                                              <td><span>99999</span><span class="ml-1">99</span><span class="ml-1">99999</span></td>
                                              <td>ACME Inc.</td>
                                              <td>01/01/2000</td>
                                              <td>
      
          <span class="badge badge-xs badge-pill badge-accent-03 text-primary">À traiter</span>
      
  
  </td>
                                              <td>
                                                  <a class="btn btn-link btn-outline-primary btn-sm" href="/approvals/prolongation/request/666">
                                                      
                                                          Traiter
                                                      
                                                  </a>
                                              </td>
                                          </tr>
                                      
                                  </tbody>
                              </table>
                          </div>
  
                          
  
  
  
                      </div>
  '''
# ---
# name: test_show_view
  '''
  <div class="c-box mb-3 mb-lg-5">
                          <h2>Détail de la demande</h2>
                          <hr/>
                          <h3>Date de fin de PASS IAE demandée : 31/01/2000</h3>
                          <p>
                              La prolongation est demandée pour une durée de 30 jours.
                          </p>
                          <h3>Motif sélectionné par l’employeur</h3>
                          <p>50 ans et plus</p>
                          
                              <h3>Bilan des actions réalisées</h3>
                              <p>
                                  Le bilan des actions réalisées et actions prévues avec le salarié, rempli par l’employeur est présent sur le document excel téléchargeable ci-dessous.
                              </p>
                              <p>
                                  <a class="btn btn-secondary btn-ico" download="Bilan prolongation PASS IAE 999999999999 John Doe.xlsx" href="https://None/None/f0r_5n4p5h07">
                                      <i class="ri-download-line ri-lg font-weight-normal"></i>
                                      <span>Télécharger le bilan</span>
                                  </a>
                              </p>
                          
                          
                              <h3>Explications supplémentaires</h3>
                              <p>
                                  L’employeur a fait une demande d'entretien téléphonique pour vous apporter des explications supplémentaires pour cette prolongation. Voici ses coordonnées :
                              </p>
                              <p>
                                  <i class="ri-mail-line"></i>
                                  <a class="btn-link mr-3" href="mailto:email@example.com">email@example.com</a>
                                  <button class="js-copy-to-clipboard btn btn-ico btn-secondary" data-copy-to-clipboard="email@example.com" data-original-title="Copié !" data-placement="top" data-toggle="tooltip" data-trigger="manual">
      <i class="ri-file-copy-line ri-lg font-weight-normal"></i>
      <span>Copier</span>
  </button>
  
                              </p>
                              <p>
                                  <i class="ri-phone-line"></i>
                                  <a class="btn-link" href="tel:+33123456789">+33123456789</a>
                              </p>
                          
                      </div>
  '''
# ---
# name: test_template_status_card[DENIED]
  '''
  <div class="c-card card rounded-lg mb-3 mb-lg-5">
      <div class="card-body">
          
              <div class="card-title">
                  <h4>Pas de prolongation du PASS IAE</h4>
              </div>
              <div class="card-text">
                  <p class="mb-2">
                      Demande refusée le 01 janvier 2000 par John Doe
                  </p>
              </div>
          
      </div>
  </div>
  
  '''
# ---
# name: test_template_status_card[GRANTED]
  '''
  <div class="c-card card rounded-lg mb-3 mb-lg-5">
      <div class="card-body">
          
              <div class="card-title">
                  <h4>Prolongation du PASS IAE jusqu’au 31/01/2000</h4>
              </div>
              <div class="card-text">
                  <p class="mb-2">
                      Demande acceptée le 01 janvier 2000 par John Doe
                  </p>
              </div>
          
      </div>
  </div>
  
  '''
# ---
# name: test_template_status_card[PENDING]
  '''
  <div class="c-card card rounded-lg mb-3 mb-lg-5">
      <div class="card-body">
          
              <div class="card-title">
                  <h4>Prolonger le PASS IAE jusqu’au 31/01/2000</h4>
              </div>
              <div class="card-text">
                  <form method="post">
                      
                      <button name="action" value="grant" class="btn btn-primary btn-block btn-ico justify-content-center">
                          <i class="ri-check-line"></i>
                          <span>Accepter</span>
                      </button>
                      <button name="action" value="deny" class="btn btn-outline-primary btn-block btn-ico mt-3 justify-content-center">
                          <i class="ri-close-line"></i>
                          <span>Refuser</span>
                      </button>
                  </form>
                  <p class="font-weight-bold mt-3">Précision</p>
                  <p class="mb-2">
                      En cas de refus, vous êtes tenu de proposer une solution au candidat.
                      Vous devez également motiver par écrit votre refus de prolongation auprès de la SIAE (Dans l'attente de la mise en œuvre du module spécifique).
                  </p>
              </div>
          
      </div>
  </div>
  
  '''
# ---
